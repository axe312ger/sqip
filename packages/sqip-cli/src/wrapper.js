#!/usr/bin/env node

import sqipCLI from './sqip-cli'
;(async () => {
  await sqipCLI()
})()
