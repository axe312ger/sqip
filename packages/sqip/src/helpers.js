import path from 'path'

import cheerio from 'cheerio'
import Debug from 'debug'
import expandTilde from 'expand-tilde'
import fastGlob from 'fast-glob'
import fs from 'fs-extra'

const debug = Debug('sqip')

export const loadSVG = (svg) => {
  return cheerio.load(svg, {
    normalizeWhitespace: true,
    xmlMode: true
  })
}

export async function locateFiles(input) {
  const enhancedInput = expandTilde(input)
  let globPattern = enhancedInput
  try {
    const stat = await fs.lstat(enhancedInput)

    if (stat.isFile()) {
      debug(`input ${input} is a file. Skip file search.`)
      return [enhancedInput]
    }

    if (stat.isDirectory()) {
      debug(
        `input ${input} is a directory. Enhancing with * to match all files.`
      )
      globPattern = `${path.resolve(enhancedInput)}/*`
    }
  } catch (err) {
    if (err.code !== 'ENOENT') {
      throw err
    }
  }

  // Find all files matching the enhanced glob
  const files = await fastGlob(globPattern, {
    onlyFiles: true,
    extglob: true,
    absolute: true
  })

  // Test if files are found
  if (!files.length) {
    throw new Error(
      `Unable to find any files via ${globPattern}. Make sure the file exists.

If you are using globbing patterns, the following features are supported:

https://github.com/micromatch/micromatch#matching-features`
    )
  }

  return files
}

export function parseColor({ palette, color }) {
  return color in palette ? palette[color].getHex() : color
}
