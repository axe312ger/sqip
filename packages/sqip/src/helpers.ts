import path from 'path'

import cheerio, { Root } from 'cheerio'
import Debug from 'debug'
import expandTilde from 'expand-tilde'
import fastGlob from 'fast-glob'
import fs from 'fs-extra'

import { Palette } from '@vibrant/color'

const debug = Debug('sqip')

export const loadSVG = (svg: string): Root => {
  return cheerio.load(svg, {
    normalizeWhitespace: true,
    xmlMode: true
  })
}

export async function locateFiles(input: string): Promise<string[]> {
  const enhancedInput = expandTilde(input)
  let globPattern = enhancedInput
  try {
    const stat = await fs.lstat(enhancedInput)

    if (stat.isFile()) {
      debug(`input ${input} is a file. Skip file search.`)
      return [enhancedInput]
    }

    if (stat.isDirectory()) {
      debug(
        `input ${input} is a directory. Enhancing with * to match all files.`
      )
      globPattern = `${path.resolve(enhancedInput)}/*`
    }
  } catch (err) {
    if (err.code !== 'ENOENT') {
      throw err
    }
  }

  // Find all files matching the enhanced glob
  const files = await fastGlob(globPattern, {
    onlyFiles: true,
    extglob: true,
    absolute: true
  })

  // Test if files are found
  if (!files.length) {
    throw new Error(
      `Unable to find any files via ${globPattern}. Make sure the file exists.

If you are using globbing patterns, the following features are supported:

https://github.com/micromatch/micromatch#matching-features`
    )
  }

  return files
}

interface ParseColorOptions {
  palette: Palette
  color: string
}

export function parseColor({ palette, color }: ParseColorOptions): string {
  // @todo test, fallback to or detect transparent as color (for bg)
  return palette[color]?.hex || color
}
