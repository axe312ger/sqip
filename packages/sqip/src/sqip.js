// #############################################################################
// #
// # "SQIP" (pronounced \skwÉªb\ like the non-magical folk of magical descent)
// # is a SVG-based LQIP technique - https://github.com/technopagan/sqip
// #
// # Installation as CLI:
// # npm install -g sqip
// #
// # Installation as lib:
// # npm install sqip
// #
// # Requirements:
// # * Node.js >= v6 (https://nodejs.org/en/)
// #
// #############################################################################

import path from 'path'

import Debug from 'debug'
import fs from 'fs-extra'

import { getDimensions, printFinalResult } from './helpers'

const debug = Debug('sqip')

export default async function sqip(options) {
  // Build configuration based on passed options and default options
  const defaultOptions = {
    plugins: [
      { name: 'primitive', options: { numberOfPrimitives: 8, mode: 0 } },
      'blur',
      'svgo',
      'data-uri'
    ],
    shouldThrow: true // @todo do we really need this?
  }
  const config = Object.assign({}, defaultOptions, options)

  // Validate configuration
  if (!config.input) {
    throw new Error(
      'Please provide an input image, e.g. sqip({ input: "input.jpg" })'
    )
  }

  const inputPath = path.resolve(process.cwd(), config.input)

  try {
    await fs.access(inputPath, fs.constants.R_OK)
  } catch (err) {
    throw new Error(`Unable to read input file: ${inputPath}`)
  }

  const dimensions = getDimensions(inputPath)

  // Resolver
  const plugins = await Promise.all(
    config.plugins.map(async plugin => {
      if (typeof plugin === 'string') {
        plugin = { name: plugin }
      }
      const { name } = plugin

      if (!name) {
        throw new Error(
          `Unable to read plugin name from:\n${JSON.stringify(plugin, null, 2)}`
        )
      }
      const moduleName =
        name.indexOf('sqip-plugin-') !== -1 ? name : `sqip-plugin-${name}`
      try {
        debug(`Loading ${moduleName}`)
        const Plugin = await import(moduleName)

        return { ...plugin, Plugin }
      } catch (err) {
        if (err.code === 'MODULE_NOT_FOUND') {
          throw new Error(
            `Unable to load plugin "${moduleName}". Try installing it via:\n\n npm install ${moduleName}`
          )
        }
        throw err
      }
    })
  )

  let svg = config.filename
  for (const { name, options: pluginOptions, Plugin } of plugins) {
    try {
      debug(`Construct ${name}`)
      const plugin = new Plugin({ dimensions, ...options, ...pluginOptions })
      debug(`Apply ${name}`)
      svg = await plugin.apply(svg, { dimensions })
    } catch (err) {
      if (config.shouldThrow) {
        throw err
      }
      console.error(err)
      process.exit(1)
    }
  }

  debug(`Finished`)

  // Write to disk or output result
  if (config.output) {
    const outputPath = path.resolve(config.output)
    // @todo remove all sync calls
    fs.writeFileSync(outputPath, svg)
  } else {
    printFinalResult(dimensions, inputPath, svg)
  }
  return { svg, dimensions }
}
