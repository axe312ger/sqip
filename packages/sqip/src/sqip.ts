import path from 'path'

import Debug from 'debug'
import fs from 'fs-extra'
import imageSize from 'probe-image-size'
import Vibrant from 'node-vibrant'
import sharp from 'sharp'
import termimg from 'term-img'
import Table from 'cli-table3'
import chalk from 'chalk'

import { Palette } from '@vibrant/color'
import { OptionDefinition } from 'command-line-args'

import { locateFiles } from './helpers'

export { loadSVG, parseColor } from './helpers'

const debug = Debug('sqip')

const mainKeys = ['originalWidth', 'originalHeight', 'width', 'height', 'type']

const PALETTE_KEYS: (keyof Palette)[] = [
  'Vibrant',
  'DarkVibrant',
  'LightVibrant',
  'Muted',
  'DarkMuted',
  'LightMuted'
]

interface SqipPluginInterface {
  metadata: SqipImageMetadata
  sqipConfig: SqipConfig
  apply(imageBuffer: Buffer): Promise<Buffer> | Buffer
}

export interface SqipResult {
  content: Buffer
  metadata: SqipImageMetadata
}

export interface SqipPluginOptions {
  pluginOptions: PluginOptions
  options: PluginOptions
  metadata: SqipImageMetadata
  sqipConfig: SqipConfig
}
export interface SqipCliOptionDefinition extends OptionDefinition {
  description?: string
  required?: boolean
  default?: boolean
}

export interface PluginOptions {
  [key: string]: unknown
}

interface PluginResolver {
  name: string
  options?: PluginOptions
}

interface SqipConfig {
  input: string | Buffer
  outputFileName?: string
  output: string
  silent?: boolean
  parseableOutput?: boolean
  plugins: PluginType[]
  width?: number
}

interface ProcessFileOptions {
  buffer: Buffer
  outputFileName: string
  config: SqipConfig
}

interface SqipImageMetadata {
  originalWidth: number
  originalHeight: number
  palette: Palette
  height: number
  width: number
  type: string
  [key: string]: unknown
}

type PluginType = PluginResolver | string

export class SqipPlugin implements SqipPluginInterface {
  public metadata: SqipImageMetadata
  public sqipConfig: SqipConfig
  public options: PluginOptions
  static cliOptions: SqipCliOptionDefinition[]

  constructor(options: SqipPluginOptions) {
    const { sqipConfig, metadata } = options
    this.sqipConfig = sqipConfig || {}
    this.metadata = metadata || {}
    this.options = {}
  }
  apply(imageBuffer: Buffer): Promise<Buffer> | Buffer {
    throw new Error('Not implemented')
    return imageBuffer
  }
}

// Resolves plugins based on a given config
// Array of plugin names or config objects, even mixed.
export async function resolvePlugins(
  plugins: PluginType[]
): Promise<(PluginResolver & { Plugin: typeof SqipPlugin })[]> {
  return Promise.all(
    plugins.map(async (plugin) => {
      if (typeof plugin === 'string') {
        plugin = { name: plugin }
      }
      const { name } = plugin

      if (!name) {
        throw new Error(
          `Unable to read plugin name from:\n${JSON.stringify(plugin, null, 2)}`
        )
      }
      const moduleName =
        name.indexOf('sqip-plugin-') !== -1 ? name : `sqip-plugin-${name}`
      try {
        debug(`Loading ${moduleName}`)
        const Plugin = await import(moduleName)

        return { ...plugin, Plugin: Plugin.default }
      } catch (err) {
        if (err.code === 'MODULE_NOT_FOUND') {
          throw new Error(
            `Unable to load plugin "${moduleName}". Try installing it via:\n\n npm install ${moduleName}`
          )
        }
        throw err
      }
    })
  )
}

async function processFile({
  buffer,
  outputFileName,
  config
}: ProcessFileOptions) {
  const { output, silent, parseableOutput } = config
  const result = await processImage({ buffer, config })
  const { content, metadata } = result
  let outputPath

  debug(`Processed ${outputFileName}`)

  // Write result svg if desired
  if (output) {
    try {
      // Test if output path already exists
      const stats = await fs.stat(output)

      // Throw if it is a file and already exists
      if (!stats.isDirectory()) {
        throw new Error(
          `File ${output} already exists. Overwriting is not yet supported.`
        )
      }
      outputPath = path.resolve(output, `${outputFileName}.svg`)
    } catch (err) {
      // Output directory or file does not exist. We will create it later on.
      outputPath = output
    }

    debug(`Writing ${outputPath}`)
    await fs.writeFile(outputPath, content)
  }

  // Gather CLI output information
  if (!silent) {
    if (outputPath) {
      console.log(`Stored at: ${outputPath}`)
    }

    // Generate preview
    if (!parseableOutput) {
      // Convert to png for image preview
      const preview = await sharp(Buffer.from(content)).png().toBuffer()

      try {
        termimg(preview, {
          fallback: () => {
            // SVG results can still be outputted as string
            if (metadata.type === 'svg') {
              console.log(content.toString())
              return
            }

            // No fallback preview solution yet for non-svg files.
            console.log(
              `Unable to render a preview for ${metadata.type} files on this machine. Try using https://iterm2.com/`
            )
          }
        })
      } catch (err) {
        if (err.name !== 'UnsupportedTerminalError') {
          throw err
        }
      }
    }

    // Metadata
    const tableConfig = parseableOutput
      ? {
          chars: {
            top: '',
            'top-mid': '',
            'top-left': '',
            'top-right': '',
            bottom: '',
            'bottom-mid': '',
            'bottom-left': '',
            'bottom-right': '',
            left: '',
            'left-mid': '',
            mid: '',
            'mid-mid': '',
            right: '',
            'right-mid': '',
            middle: ' '
          },
          style: { 'padding-left': 0, 'padding-right': 0 }
        }
      : undefined

    // Figure out which metadata keys to show
    // @todo why is this unused?
    // const allKeys = [...mainKeys, 'palette']

    const mainTable = new Table(tableConfig)
    mainTable.push(mainKeys)
    mainTable.push(
      mainKeys.map((key) => String(metadata[key]) || 'can not display')
    )
    console.log(mainTable.toString())

    // Show color palette
    const paletteTable = new Table(tableConfig)
    paletteTable.push(PALETTE_KEYS)
    paletteTable.push(
      PALETTE_KEYS.map((key) => metadata.palette[key]?.hex)
        .filter<string>((hex): hex is string => typeof hex === 'string')
        .map((hex) => chalk.hex(hex)(hex))
    )
    console.log(paletteTable.toString())

    Object.keys(metadata)
      .filter((key) => ![...mainKeys, 'palette'].includes(key))
      .forEach((key) => {
        console.log(chalk.bold(`${key}:`))
        console.log(metadata[key])
      })
  }

  return result
}

interface ProcessImageOptions {
  buffer: Buffer
  config: SqipConfig
}

async function processImage({
  buffer,
  config
}: ProcessImageOptions): Promise<SqipResult> {
  const originalSizes = imageSize.sync(buffer)
  const vibrant = Vibrant.from(buffer)
  const palette = await vibrant.quality(0).getPalette()

  if (!originalSizes) {
    throw new Error('Unable to get image size')
  }

  const metadata: SqipImageMetadata = {
    originalWidth: originalSizes.width,
    originalHeight: originalSizes.height,
    palette,
    type: 'unknown',
    width: 0,
    height: 0
  }

  // Load plugins
  const plugins = await resolvePlugins(config.plugins)

  // Interate through plugins and apply them to last returned image
  if (config.width && config.width > 0) {
    // Resize to desired output width
    buffer = await sharp(buffer).resize(config.width).toBuffer()

    const resizedMetadata = await sharp(buffer).metadata()
    metadata.width = resizedMetadata.width || 0
    metadata.height = resizedMetadata.height || 0
  } else {
    // Fall back to original size, keep image as is
    metadata.width = originalSizes.width
    metadata.height = originalSizes.height
  }

  for (const { name, options: pluginOptions, Plugin } of plugins) {
    debug(`Construct ${name}`)
    const plugin = new Plugin({
      sqipConfig: config,
      pluginOptions: pluginOptions || {},
      metadata,
      options: {}
    })
    debug(`Apply ${name}`)
    buffer = await plugin.apply(buffer)
  }

  return { content: buffer, metadata }
}

export default async function sqip(
  options: SqipConfig
): Promise<SqipResult | SqipResult[]> {
  // Build configuration based on passed options and default options
  const defaultOptions = {
    plugins: [
      { name: 'primitive', options: { numberOfPrimitives: 8, mode: 0 } },
      'blur',
      'svgo',
      'data-uri'
    ],
    width: 300,
    parseableOutput: false,
    silent: true
  }

  const config = Object.assign({}, defaultOptions, options)

  const { input, outputFileName, parseableOutput, silent } = config

  if (parseableOutput) {
    chalk.level = 0
  }

  // Validate configuration
  if (!input) {
    throw new Error(
      'Please provide an input image, e.g. sqip({ input: "input.jpg" })'
    )
  }

  // If input is a Buffer
  if (Buffer.isBuffer(input)) {
    if (!outputFileName) {
      throw new Error(
        `${outputFileName} is required when passing image as buffer`
      )
    }
    return processFile({
      buffer: input,
      outputFileName,
      config
    })
  }

  const files = await locateFiles(input)
  debug('Found files:')
  debug(files)

  // Test if all files are accessable
  for (const file of files) {
    try {
      debug('check file ' + file)
      await fs.access(file, fs.constants.R_OK)
    } catch (err) {
      throw new Error(`Unable to read file ${file}`)
    }
  }

  // Iterate over all files
  const results = []
  for (const filePath of files) {
    // Apply plugins to files
    if (!silent) {
      console.log(`Processing: ${filePath}`)
    } else {
      debug(`Processing ${filePath}`)
    }
    const buffer = await fs.readFile(filePath)
    const result = await processFile({
      buffer,
      outputFileName: outputFileName || path.parse(filePath).name,
      config
    })

    results.push(result)
  }

  debug(`Finished`)

  // Return as array when input was array or results is only one file
  if (Array.isArray(input) || results.length === 0) {
    return results
  }
  return results[0]
}

export * from './helpers'
