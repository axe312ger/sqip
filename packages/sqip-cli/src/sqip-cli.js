import Debug from 'debug'
import commandLineArgs from 'command-line-args'
import commandLineUsage from 'command-line-usage'

import sqip, { resolvePlugins } from 'sqip'

const { version } = require('../package.json')

const debug = Debug('sqip-cli')

let optionList = [
  {
    name: 'help',
    alias: 'h',
    description: 'Show help'
  },
  {
    name: 'version',
    description: 'Show version number'
  },
  {
    name: 'plugins',
    alias: 'p',
    type: String,
    multiple: true,
    description: 'One or more plugins. E.g. "-p primitive blur"'
  },
  {
    name: 'input',
    alias: 'i',
    type: String,
    required: true
  },
  {
    name: 'output',
    alias: 'o',
    type: String,
    description:
      'Save the resulting SVG to a file. The svg result will be returned by default.'
  }
]

function showHelp({ optionList }) {
  const sections = [
    {
      header: 'sqip CLI',
      content:
        'Usage: sqip --input [path]\n\n"SQIP" (pronounced \\skwÉªb\\ like the non-magical folk of magical descent) is a SVG-based LQIP technique - https://github.com/technopagan/sqip'
    },
    {
      header: 'Options',
      optionList
    },
    {
      header: 'Examples',
      content: `Output input.jpg image as SQIP
$ sqip --input /path/to/input.jpg

Save input.jpg as result.svg with 25 shapes and no blur
$ sqip -i input.jpg -n 25 -b 0 -o result.svg`
    }
  ]
  const usage = commandLineUsage(sections)
  console.log(usage)
}

const args = commandLineArgs(optionList, { partial: true })

if ('version' in args) {
  console.log(version)
  process.exit(0)
}

let { plugins } = args

// Default cli plugins
if (!plugins) {
  plugins = ['primitive', 'blur', 'svgo'].filter(Boolean)
}

;(async () => {
  debug(`Found plugins:\n`, plugins)
  const resolvedPlugins = await resolvePlugins(plugins)

  // Add new cli options based on enabled plugins
  const pluginOptions = resolvedPlugins.reduce((definitions, plugin) => {
    const {
      name,
      Plugin: { cliOptions }
    } = plugin
    if (cliOptions) {
      return [
        ...definitions,
        ...cliOptions.map(option => ({
          ...option,
          name: `${name}-${option.name}`
        }))
      ]
    }
    return definitions
  }, [])

  optionList = [...optionList, ...pluginOptions]

  debug(`Generated CLI options:\n`, optionList)

  const args = commandLineArgs(optionList, { partial: true })

  debug(`Parsed CLI args:\n`, args)

  if ('help' in args) {
    showHelp({ optionList })
    process.exit(0)
  }

  const missing = optionList
    .filter(({ required }) => required)
    .filter(({ name }) => !args[name])
    .map(({ name }) => name)

  if (missing.length) {
    showHelp({ optionList })
    console.info(
      `\nPlease provide the following arguments: ${missing.join(', ')}`
    )
    process.exit(1)
  }

  const { input, output } = args

  // Build list of plugins with options based on passed arguments
  const pluginsOptions = resolvedPlugins.map(({ name }) => {
    const options = Object.keys(args)
      .filter(args => args.indexOf(`${name}-`) === 0)
      .reduce((optionMap, argName) => {
        const optionName = argName.substr(`${name}-`.length)
        return { ...optionMap, [optionName]: args[argName] }
      }, {})
    return { name, options }
  })

  const options = {
    input,
    output,
    plugins: pluginsOptions
  }

  try {
    await sqip(options)
  } catch (err) {
    console.log(err)
    process.exit(1)
  }
})()
