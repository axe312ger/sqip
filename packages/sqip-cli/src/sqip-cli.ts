import path from 'path'

import commandLineArgs from 'command-line-args'
import commandLineUsage from 'command-line-usage'
import Debug from 'debug'
import fs from 'fs-extra'
import pkgUp from 'pkg-up'

import sqip, { resolvePlugins, SqipCliOptionDefinition } from 'sqip'

const debug = Debug('sqip-cli')

const defaultOptionList: SqipCliOptionDefinition[] = [
  {
    name: 'help',
    alias: 'h',
    description: 'Show help'
  },
  {
    name: 'version',
    description: 'Show version number'
  },
  {
    name: 'plugins',
    alias: 'p',
    type: String,
    multiple: true,
    description: 'One or more plugins. E.g. "-p primitive blur"'
  },
  {
    name: 'input',
    alias: 'i',
    type: String,
    required: true
  },
  {
    name: 'output',
    alias: 'o',
    type: String,
    description:
      'Define the path of the resulting file. By default SQIP will guess the output file name.'
  },
  {
    name: 'width',
    alias: 'w',
    type: Number,
    defaultValue: 300,
    description:
      'Width of the resulting file. Negative values and 0 will fall back to original image width.'
  },
  {
    name: 'silent',
    type: Boolean,
    defaultValue: false,
    description: 'Supress all output'
  },
  {
    name: 'parseable-output',
    type: Boolean,
    defaultValue: false,
    description:
      'Ensure the output is parseable. Will suppress the preview images and the table borders.'
  }
]

function showHelp({ optionList }: { optionList: SqipCliOptionDefinition[] }) {
  const sections = [
    {
      header: 'sqip CLI',
      content:
        'Usage: sqip --input [path]\n\n"SQIP" (pronounced \\skwÉªb\\ like the non-magical folk of magical descent) is a SVG-based LQIP technique - https://github.com/technopagan/sqip'
    },
    {
      header: 'Options',
      optionList
    },
    {
      header: 'Examples',
      content: `Output input.jpg image as SQIP
$ sqip --input /path/to/input.jpg

Save input.jpg as result.svg with 25 shapes and no blur
$ sqip -i input.jpg -n 25 -b 0 -o result.svg`
    }
  ]
  const usage = commandLineUsage(sections)
  console.log(usage)
}

export default async function sqipCLI(): Promise<undefined> {
  const pluginDetectionArgs = commandLineArgs(defaultOptionList, {
    partial: true
  })

  if ('version' in pluginDetectionArgs) {
    const closestPackageJSON = await pkgUp()
    if (!closestPackageJSON) {
      throw new Error('Unable to detect CLI version')
    }
    const { version } = await fs.readJSON(closestPackageJSON)
    console.log(version)
    return process.exit(0)
  }

  let { plugins } = pluginDetectionArgs

  // Default cli plugins
  if (!plugins) {
    plugins = ['primitive', 'blur', 'svgo'].filter(Boolean)
  }

  debug(`Found plugins:\n`, plugins)
  let resolvedPlugins

  try {
    resolvedPlugins = await resolvePlugins(plugins)
  } catch (err) {
    console.error(err)
    process.exit(1)
  }

  // Add new cli options based on enabled plugins
  const pluginOptions = resolvedPlugins.reduce<SqipCliOptionDefinition[]>(
    (definitions, plugin) => {
      const {
        name,
        Plugin: { cliOptions }
      } = plugin
      if (cliOptions) {
        return [
          ...definitions,
          ...cliOptions.map((option) => ({
            ...option,
            name: `${name}-${option.name}`
          }))
        ]
      }
      return definitions
    },
    []
  )

  const optionList = [...defaultOptionList, ...pluginOptions]

  debug(`Generated CLI options:\n`, optionList)

  const args = commandLineArgs(optionList, { partial: true })

  debug(`Parsed CLI args:\n`, args)

  if ('help' in args) {
    showHelp({ optionList })
    return process.exit(0)
  }

  const missing = optionList
    .filter(({ required }) => required)
    .filter(({ name }) => !args[name])
    .map(({ name }) => name)

  if (missing.length) {
    showHelp({ optionList })
    console.error(
      `\nPlease provide the following arguments: ${missing.join(', ')}`
    )
    return process.exit(1)
  }

  const { input, output, width } = args
  const { name } = path.parse(input)
  const guessedOutput = path.resolve(process.cwd(), `${name}.svg`)

  // Build list of plugins with options based on passed arguments
  const pluginsOptions = resolvedPlugins.map(({ name }) => {
    const options = Object.keys(args)
      .filter((args) => args.indexOf(`${name}-`) === 0)
      .reduce((optionMap, argName) => {
        const optionName = argName.substr(`${name}-`.length)
        return { ...optionMap, [optionName]: args[argName] }
      }, {})
    return { name, options }
  })

  const options = {
    input,
    output: output || guessedOutput,
    width,
    plugins: pluginsOptions,
    silent: args.silent,
    parseableOutput: args['parseable-output']
  }

  debug(`Final sqip options:`, options)

  try {
    await sqip(options)
  } catch (err) {
    console.error(err)
    process.exit(1)
  }
}
