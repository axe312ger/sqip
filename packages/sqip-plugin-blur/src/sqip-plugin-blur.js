import { loadSVG, SqipPlugin } from 'sqip'

const PRIMITIVE_SVG_ELEMENTS = 'circle, ellipse, line, polygon, path, rect, g'

const patchSVGGroup = svg => {
  const $ = loadSVG(svg)

  const $svg = $('svg')
  const $primitiveShapes = $svg.children(PRIMITIVE_SVG_ELEMENTS)

  // Check if actual shapes are grouped
  if (!$primitiveShapes.filter('g').length !== 1) {
    const $group = $('<g/>')
    const $realShapes = $primitiveShapes.not('rect:first-child')

    $group.append($realShapes)
    $svg.append($group)
  }

  return $.html()
}

export default class SVGPlugin extends SqipPlugin {
  static get cliOptions() {
    return [
      {
        name: 'blur',
        alias: 'b',
        type: Number,
        description:
          'Set the stdDeviation value for the GaussianBlur SVG filter. See: https://developer.mozilla.org/en-US/docs/Web/SVG/Element/feGaussianBlur',
        defaultValue: 55
      }
    ]
  }
  constructor({ pluginOptions }) {
    super(...arguments)
    this.options = { blur: 55, ...pluginOptions }
  }

  apply(svg) {
    svg = this.prepareSVG(svg)
    if (this.options.blur) {
      svg = this.applyBlurFilter(svg)
    }
    return svg
  }

  // Prepare SVG. For now, this will just ensure that the viewbox attribute is set
  prepareSVG(svg) {
    const $ = loadSVG(svg)
    const $svg = $('svg')
    const { width, height } = this.metadata

    // Ensure viewbox
    if (!$svg.is('[viewBox]')) {
      if (!(width && height)) {
        throw new Error(
          `SVG is missing viewBox attribute while Width and height were not passed:\n\n${svg}`
        )
      }
      $svg.attr('viewBox', `0 0 ${width} ${height}`)
    }

    const $bgRect = $svg
      .children(PRIMITIVE_SVG_ELEMENTS)
      .filter('rect:first-child[fill]')

    // Check if filling background rectangle exists
    // This must exist for proper blur and other transformations
    if (!$bgRect.length) {
      throw new Error(
        `The SVG must have a rect as first shape element which represents the svg background color:\n\n${svg}`
      )
    }

    // Remove x and y attributes since they default to 0
    // @todo test in rare browsers
    $bgRect.attr('x', null)
    $bgRect.attr('y', null)

    // Improve compression via simplifying fill
    $bgRect.attr('width', '100%')
    $bgRect.attr('height', '100%')

    return $.html()
  }

  applyBlurFilter(svg) {
    if (!this.options.blur) {
      return svg
    }
    const patchedSVG = patchSVGGroup(svg)
    const $ = loadSVG(patchedSVG)
    const blurFilterId = 'b'
    $('svg > g').attr('filter', `url(#${blurFilterId})`)
    $('svg').prepend(
      `<filter id="${blurFilterId}"><feGaussianBlur stdDeviation="${
        this.options.blur
      }" />`
    )

    return $.html()
  }
}
