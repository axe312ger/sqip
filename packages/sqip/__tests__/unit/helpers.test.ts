import { resolve } from 'path'
import { readFile } from 'fs-extra'

import { loadSVG, locateFiles } from '../../src/helpers'

const DIR_ROOT = resolve(__dirname, '../../../..')
const DIR_FIXTURES = resolve(DIR_ROOT, '__tests__', 'fixtures')

const FILE_JPG = resolve(DIR_FIXTURES, 'beach.jpg')
const FILE_SVG = resolve(DIR_FIXTURES, 'beach-sqip.svg')

test('loadSVG', async () => {
  const svgContentBuffer = await readFile(FILE_SVG)
  const $svg = loadSVG(svgContentBuffer.toString())
  expect($svg('svg')).toHaveLength(1)
})

const cleanResultArray = (results: string[]) =>
  results.map((result) => result.replace(DIR_ROOT, ''))

describe('locateFiles', () => {
  test('invalid directory or files yields empty array', async () => {
    await expect(
      locateFiles('/foo/bar/baz')
    ).rejects.toThrowErrorMatchingSnapshot()
  })
  test('invalid glob', async () => {
    await expect(
      locateFiles('/foo/bar/baz/*')
    ).rejects.toThrowErrorMatchingSnapshot()
  })
  test('single file', async () => {
    const result = await locateFiles(FILE_JPG)
    expect(result).toHaveLength(1)
    expect(cleanResultArray(result)).toMatchSnapshot()
  })
  test('whole directory without glob', async () => {
    const result = await locateFiles(DIR_FIXTURES)
    expect(result).toHaveLength(3)
    expect(cleanResultArray(result)).toMatchSnapshot()
  })
  test('whole directory with glob "*"', async () => {
    const result = await locateFiles(`${DIR_FIXTURES}/*`)
    expect(result).toHaveLength(3)
    expect(cleanResultArray(result)).toMatchSnapshot()
  })
  test('whole directory with glob "*.{jpg,jpeg}"', async () => {
    const result = await locateFiles(`${DIR_FIXTURES}/*.{jpg,jpeg}`)
    expect(result).toHaveLength(2)
    expect(cleanResultArray(result)).toMatchSnapshot()
  })
  test('subdirectory glob "**/*"', async () => {
    const result = await locateFiles(`${resolve(DIR_FIXTURES, '..')}/**/*`)
    expect(result).toHaveLength(3)
    expect(cleanResultArray(result)).toMatchSnapshot()
  })
  test('expandes tilde in glob "~+/__tests__/fixtures"', async () => {
    const result = await locateFiles(`~+/__tests__/fixtures`)
    expect(result).toHaveLength(3)
    expect(cleanResultArray(result)).toMatchSnapshot()
  })
})
