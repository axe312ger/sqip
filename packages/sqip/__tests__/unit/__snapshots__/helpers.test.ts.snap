// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`locateFiles expandes tilde in glob "~+/__tests__/fixtures" 1`] = `
Array [
  "/__tests__/fixtures/CMYK-example.jpg",
  "/__tests__/fixtures/beach-sqip.svg",
  "/__tests__/fixtures/beach.jpg",
]
`;

exports[`locateFiles invalid directory or files yields empty array 1`] = `
"Unable to find any files via /foo/bar/baz. Make sure the file exists.

If you are using globbing patterns, the following features are supported:

https://github.com/micromatch/micromatch#matching-features"
`;

exports[`locateFiles invalid glob 1`] = `
"Unable to find any files via /foo/bar/baz/*. Make sure the file exists.

If you are using globbing patterns, the following features are supported:

https://github.com/micromatch/micromatch#matching-features"
`;

exports[`locateFiles single file 1`] = `
Array [
  "/__tests__/fixtures/beach.jpg",
]
`;

exports[`locateFiles subdirectory glob "**/*" 1`] = `
Array [
  "/__tests__/fixtures/CMYK-example.jpg",
  "/__tests__/fixtures/beach-sqip.svg",
  "/__tests__/fixtures/beach.jpg",
]
`;

exports[`locateFiles whole directory with glob "*" 1`] = `
Array [
  "/__tests__/fixtures/CMYK-example.jpg",
  "/__tests__/fixtures/beach-sqip.svg",
  "/__tests__/fixtures/beach.jpg",
]
`;

exports[`locateFiles whole directory with glob "*.{jpg,jpeg}" 1`] = `
Array [
  "/__tests__/fixtures/CMYK-example.jpg",
  "/__tests__/fixtures/beach.jpg",
]
`;

exports[`locateFiles whole directory without glob 1`] = `
Array [
  "/__tests__/fixtures/CMYK-example.jpg",
  "/__tests__/fixtures/beach-sqip.svg",
  "/__tests__/fixtures/beach.jpg",
]
`;
